vnext/Microsoft.ReactNative/Utils/TextHitTestUtils.cpp

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+
+#include "TextHitTestUtils.h"
+
+#include <UI.Xaml.Documents.h>
+
+namespace Microsoft::ReactNative {
+
+static bool IsPointAfterCharacter(
+    const winrt::Point &point,
+    const xaml::Documents::TextPointer &textPointer,
+    const winrt::Rect &rect,
+    bool isRtl,
+    double endOfLine) {
+  // The point is after the character if below the bottom of the character or
+  // if on the same line and "after" the end of the character (greater than
+  // the X-dimension of the next character rect in LTR, less than for RTL):
+  // ┌────> X
+  // │ ┌───────────┐  ┌───────────┐
+  // ↓ │     ┌─────┘  └─────┐     │
+  // Y └─────┘ (x,y)  (x,y) └─────┘
+  //        LTR            RTL
+  if (point.Y < rect.Y) {
+    // Point is above the top of the character.
+    return false;
+  }
+
+  const auto bottom = rect.Y + rect.Height;
+  if (point.Y > bottom) {
+    // Point is below the bottom of the character.
+    return true;
+  }
+
+  auto end = endOfLine;
+  const auto endPointer = textPointer.GetPositionAtOffset(1, xaml::Documents::LogicalDirection::Forward);
+  if (endPointer != nullptr) {
+    const auto endRect = endPointer.GetCharacterRect(xaml::Documents::LogicalDirection::Forward);
+    if (rect.Y == endRect.Y) {
+      end = endRect.X;
+    }
+  }
+
+  return !isRtl ? point.X >= end : point.X <= end;
+}
+
+static bool IsPointBeforeCharacter(
+    const winrt::Point &point,
+    const xaml::Documents::TextPointer &textPointer,
+    winrt::Rect rect,
+    bool isRtl) {
+  // The point is before the character if above the top of the character or if
+  // on the same line and "before" the start of the character (less than the
+  // X-dimension of the character rect in LTR, greater than for RTL):
+  // ┌────> X
+  // │ (x,y) ┌─────┐  ┌─────┐ (x,y)
+  // ↓ ┌─────┘     │  │     └─────┐
+  // Y └───────────┘  └───────────┘
+  //        LTR            RTL
+  if (point.Y < rect.Y) {
+    // Point is above the top of the character.
+    return true;
+  }
+
+  const auto bottom = rect.Y + rect.Height;
+  if (point.Y > bottom) {
+    // Point is below the bottom of the character.
+    return false;
+  }
+
+  // The point is on the same line, check if the point is before the start of
+  // the character.
+  return !isRtl ? point.X < rect.X : point.X > rect.X;
+}
+
+static bool IsRTL(const xaml::Documents::TextPointer &textPointer) {
+  auto currentPointer = textPointer.GetPositionAtOffset(1, xaml::Documents::LogicalDirection::Forward);
+  auto currentRect = textPointer.GetCharacterRect(xaml::Documents::LogicalDirection::Forward);
+  auto firstCharacterRect = currentRect;
+  while (currentPointer != nullptr) {
```

Comment:
What's the worst-case time complexity of this function?

Can you get the value of the isRtl prop from the session?

Reply:
Worst case complexity is O(log(c)), where 'c' is the number of characters in the Run. Please note though that by default this penalty is only paid on runs that are both:
1. Pressable
2. Multiline

So the cost is only incurred when someone adds a pressable Span that wraps or has line breaks.

Reply:
Actually, if you want to get specific, worst case is O(S * log(c)) where S is the number of pressable spans that are multiline and c is average length of these spans.

Reply:
And for comparison, the current algorithm is O(N), where N is the total number of inlines in the text sub-tree.

So take for example a tree with 10 inlines, one of which is a pressable (S*) span at depth 2 on a single line with 16 characters in it's run:
```
        T
   /    |    \
  S     S     S
 /\    /\      \
R  R  R  R      S*
               /
              R 
```
Previous algorithm: O(N) = O(10)
New algorithm: O(P + (R * log(C))) = O(4 + (0 * log(16)) = O(4)

If instead the run with 16 characters is multiline:
Previous algorithm: No change, but hit testing does not work correctly
New algorithm: O(P + (R * log(C))) = O(4 + (1 * log(16)) = O(8)

Where this starts to break down is when you have pressable spans with many nested multiline runs. I'm not sure under what conditions we might see this, but this is why I also want to propose a new prop called `hitTestStrategy`, which would allow us to bail out of recursion and instead switch directly to binary search so we pay the logarithmic penalty only once for the pressable

Reply:
I'd argue that most text controls will be a single run with a bunch of text.
Looking at a random resource file with a bunch of strings, I see that the average is about 17 characters per text string:

```ps
PS C:\> gc en.rc | %{$_.split('"')[1]} | Where {-not [string]::IsNullOrWhiteSpace($_)  } | Measure-Object -Property length -Average

Count             : 174
Average           : 16.6666666666667
Sum               :
Maximum           :
Minimum           :
StandardDeviation :
Property          : Length
```

Reply:
@asklar if none of those are strings are pressable or multiline, then the behavior doesn't change.

Reply:
I would guess that most pressable nested strings are either a) links, which are on average 62 characters according to the internet (though I'd guess the average is shorter as there is probably a preference to shorten longer URLs when presenting for human readability), or button labels / anchor text, which is generally also short.

Reply:
That's a pretty common case. For example Hebrew is rtl for text but ltr for numbers.

Reply:
@rectified95 sorry, this algorithm is O(1). Typically you only need to detect 1-2 characters to detect LTR vs. RTL. There are some characters with 0 width (like line breaks) that force us to check more than 1 character ahead, but generally speaking this is detected on the 2nd character of the Run.

Reply:
@asklar - I explicitly called out in the assumptions that we are not supporting bidirectional text at this time. One work around is to make sure you put LTR and RTL text in separate runs to get hit testing to work appropriately. Unfortunately there is not a good algorithm for bidirectional hit testing that is also logarithmic.

Reply:
@asklar to be fair, the current algorithm also does not support bidirectional text, so this is still a net improvement.